ext/bcmath/bcmath.c

```diff
@@ -152,7 +152,11 @@ PHP_FUNCTION(bcadd)
 	ZEND_PARSE_PARAMETERS_END();
 
 	if (ZEND_NUM_ARGS() == 3) {
-		scale = (int) (scale_param < 0 ? 0 : scale_param);
+	    if ((int)scale_param < 0) {
```

Comment:
In my opinion, this is just wrong (the former code either). Casting a `zend_long` to `int` causes truncation on 64bit architectures (according to the C standard it is implementation defined behavior, but I think we can assume that all supported architectures use two's complement integer representation, and that they implement the same truncation behavior). Anyway, that is certainly unexpected for PHP scripts. Instead we should check not only that `scale_param` is greater than or equal to zero, but also that it is less than or equal to `INT_MAX`. If that fails, throw an error; if the checck succeeds, cast to `int` (but not before).

Basically the same check should also be done whenever `bcmath.scale` is set.

Reply:
@cmb69 Done. Is that a correct implementation?

Can you also explain me, why should I compare with `INT_MAX`? And not with `PHP_INT_MAX` (which is `ZEND_LONG_MAX` internally)?

Reply:
Yes, the implementation looks good to me now (I took only a quick glance though).

The check for `INT_MAX` is necessary, because libbcmath declares scale as `int`, not as `zend_long`. Since we have that library bundled, we could change that, but I don't see a point in having scale factors which would exceed `INT_MAX`; such numbers would not fit in memory anyway.